pkg/routeagent_driver/handlers/calico/ippool_handler.go

```diff
@@ -0,0 +1,198 @@
+
+	return errors.Wrap(err, "error initializing Calico clientset")
+}
+
+func (h *calicoIPPoolHandler) RemoteEndpointCreated(endpoint *submV1.Endpoint) error {
+	h.remoteEndpoints[endpoint.Name] = endpoint
+	if !h.isGateway {
+		logger.V(log.TRACE).Info("Ignore RemoteEndpointCreated event (node isn't Gateway)")
+		return nil
+	}
+
+	err := h.createIPPool(endpoint)
+
+	return errors.Wrap(err, "failed to handle RemoteEndpointCreated event")
+}
+
+func (h *calicoIPPoolHandler) RemoteEndpointRemoved(endpoint *submV1.Endpoint) error {
+	delete(h.remoteEndpoints, endpoint.Name)
+
+	if !h.isGateway {
+		logger.V(log.TRACE).Info("Ignore RemoteEndpointRemoved event (node isn't Gateway)")
+		return nil
+	}
+
+	err := h.deleteIPPool(endpoint)
+
+	return errors.Wrap(err, "failed to handle RemoteEndpointRemoved event")
+}
+
+func (h *calicoIPPoolHandler) TransitionToNonGateway() error {
+	logger.Info("TransitionToNonGateway")
+
+	h.isGateway = false
+
+	return nil
+}
+
+func (h *calicoIPPoolHandler) TransitionToGateway() error {
+	logger.Info("TransitionToGateway")
+
+	retErr := error(nil)
+	h.isGateway = true
+
+	for _, endpoint := range h.remoteEndpoints {
+		err := h.createIPPool(endpoint)
+		if err != nil {
+			logger.Warningf("Failed to create ippool %s", endpoint.GetName())
+
+			retErr = err
+		}
+	}
+
+	return retErr
+}
+
+func (h *calicoIPPoolHandler) Stop(uninstall bool) error {
+	if !uninstall || !h.isGateway {
+		return nil
+	}
+
+	logger.Info("Uninstalling Calico IPPools used for Submariner")
+
+	labelSelector := labels.SelectorFromSet(map[string]string{submarinerIPPool: "true"}).String()
+	err := h.client.ProjectcalicoV3().IPPools().DeleteCollection(context.TODO(), metav1.DeleteOptions{},
+		metav1.ListOptions{LabelSelector: labelSelector})
+
+	if err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrapf(err, "Failed to delete Calico IPPools using labelSelector %q", labelSelector)
+	}
+
+	logger.Infof("Successfully delete Calico IPPools using labelSelector %q", labelSelector)
+
+	return nil
+}
+
+func (h *calicoIPPoolHandler) createIPPool(endpoint *submV1.Endpoint) error {
+	subnets := cidr.ExtractIPv4Subnets(endpoint.Spec.Subnets)
+	retErr := error(nil)
+
+	for _, subnet := range subnets {
+		iPPoolObj := &calicoapi.IPPool{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   getEndpointSubnetIPPoolName(endpoint, subnet),
+				Labels: map[string]string{submarinerIPPool: "true"},
+			},
+			Spec: calicoapi.IPPoolSpec{
+				CIDR:        subnet,
+				NATOutgoing: false,
+				Disabled:    true,
+			},
+		}
+
+		if _, err := h.client.ProjectcalicoV3().IPPools().Create(context.TODO(), iPPoolObj,
+			metav1.CreateOptions{}); err != nil && !apierrors.IsAlreadyExists(err) {
+			logger.Errorf(err, "Error creating Calico IPPool for ClusterID %q subnet %q", endpoint.Spec.ClusterID, subnet)
+			retErr = err
+		} else {
+			logger.Infof("Successfully create Calico IPPool %q", iPPoolObj.GetName())
+		}
+	}
```

Comment:
Minor thing but we should only log if it was actually created. Also we don't need to log the error here since we're returning it:

```suggestion
		err := h.client.ProjectcalicoV3().IPPools().Create(context.TODO(), iPPoolObj,
			metav1.CreateOptions{}))
	         if err == nil {
	             logger.Infof("Successfully created Calico IPPool %q", iPPoolObj.GetName())
	             continue
	         }
	         
	         if !apierrors.IsAlreadyExists(err) {
			retErrors = append(retErrors, errors.Wrapf(err, "error creating Calico IPPool for ClusterID %q subnet %q", endpoint.Spec.ClusterID, subnet)))
			
		}
	}
```